pkg/services/govmomi/vcenter/clone.go

```diff
@@ -190,6 +190,40 @@ func Clone(ctx *context.VMContext, bootstrapData []byte) error {
 		spec.Location.Datastore = datastoreRef
 	}
 
+	if ctx.VSphereVM.Spec.DatastoreCluster != "" {
+		// verify the datastoreCluster exists
+		datastoreCluster, err := ctx.Session.Finder.DatastoreCluster(ctx, ctx.VSphereVM.Spec.DatastoreCluster)
+		if err != nil {
+			return errors.Wrapf(err, "unable to get datastoreCluster %s for %v", ctx.VSphereVM.Spec.DatastoreCluster, ctx)
+		}
+		// if only the datastore cluster is mentioned
+		if datastoreRef == nil {
+			rs, _ := datastoreCluster.Children(ctx)
+			rand.Seed(time.Now().UnixNano())
+			datastoreRef = types.NewReference(rs[rand.Intn(len(rs))].Reference())
+			spec.Location.Datastore = datastoreRef
+		} else {
+			// if the datastore is also defined by the user
+			// verify that the datastore is a member of the datastoreCluster
+			memberDatastoreFound := false
+			memberDatastoresRef, _ := datastoreCluster.Children(ctx)
+			for _, datastore := range memberDatastoresRef {
+				datastoreName, err := object.NewDatastore(ctx.Session.Client.Client, datastore.Reference()).ObjectName(ctx)
+				if err != nil {
+					return errors.Wrapf(err, "unable to fetch datastore name for datastoreCluster", ctx.VSphereVM.Spec.DatastoreCluster)
+				}
+				if ctx.VSphereVM.Spec.Datastore == datastoreName {
+					memberDatastoreFound = true
+					spec.Location.Datastore = types.NewReference(datastore.Reference())
+					break
+				}
+			}
+			if !memberDatastoreFound {
+				return errors.New(fmt.Sprintf("unable to find specified datastore: %s in member list of datastores for storageCluster", ctx.VSphereVM.Spec.Datastore))
```

Comment:
it would be good if these errors were captured in a condition. Maybe move the datastoreCluster reconciliation into a separate function and then capture any errors into a condition.